src/details/ArborX_DetailsTreeTraversal.hpp

```diff
@@ -273,75 +273,97 @@ struct TreeTraversal<BVH, Predicates, Callback, NearestPredicateTag>
+      bool traverse_left = false;
+      bool traverse_right = false;
+
+      if (distance_node < radius)
       {
-        if (node->isLeaf())
+        // Insert children into the stack and make sure that the
+        // closest one ends on top.
+        child_left = bvh_.getNodePtr(node->children.first);
+        child_right = bvh_.getNodePtr(node->children.second);
+
+        distance_left = distance(child_left);
+        distance_right = distance(child_right);
+
+        if (distance_left < radius && child_left->isLeaf())
         {
-          int const leaf_index = node->getLeafPermutationIndex();
-          auto const leaf_distance = node_distance;
+          auto leaf_pair = Kokkos::make_pair(
+              child_left->getLeafPermutationIndex(), distance_left);
           if ((int)heap.size() < k)
-          {
-            // Insert leaf node and update radius if it was the kth
-            // one.
-            heap.push(Kokkos::make_pair(leaf_index, leaf_distance));
-            if ((int)heap.size() == k)
-              radius = heap.top().second;
-          }
+            heap.push(leaf_pair);
           else
-          {
-            // Replace top element in the heap and update radius.
-            heap.popPush(Kokkos::make_pair(leaf_index, leaf_distance));
+            heap.popPush(leaf_pair);
+          if ((int)heap.size() == k)
             radius = heap.top().second;
-          }
-          auto const &pair = *--stack_ptr;
-          node = pair.first;
-          node_distance = pair.second;
         }
-        else
+
+        // Note: radius may have been already updated here from the left child
+        if (distance_right < radius && child_right->isLeaf())
         {
-          // Insert children into the stack and make sure that the
-          // closest one ends on top.
-          Node const *left_child = bvh_.getNodePtr(node->children.first);
-          Node const *right_child = bvh_.getNodePtr(node->children.second);
-          auto const left_child_distance = distance(left_child);
-          auto const right_child_distance = distance(right_child);
-          if (left_child_distance < right_child_distance)
-          {
-            // NOTE not really sure why but it performed better with
-            // the conditional insertion on the device and without
-            // it on the host (~5% improvement for both)
-#if defined(__CUDA_ARCH__)
-            if (right_child_distance < radius)
-#endif
-              *stack_ptr++ = {right_child, right_child_distance};
-            node = left_child;
-            node_distance = left_child_distance;
-          }
+          auto leaf_pair = Kokkos::make_pair(
+              child_right->getLeafPermutationIndex(), distance_right);
+          if ((int)heap.size() < k)
+            heap.push(leaf_pair);
           else
-          {
+            heap.popPush(leaf_pair);
+          if ((int)heap.size() == k)
+            radius = heap.top().second;
+        }
```

Comment:
The duplication of code is unfortunate.

Reply:
Yeah, I was really annoyed by the duplication. The only way to clean it up would be to to move part of the logic inside heap itself, but it's another can of worms.